File: PostProcessing/MergePPKGPS_sonar.py
"""Converted from MergePPKGPS_sonar.m by Spicer Bak and Chat GPT conversion on 3 July 2023.
This takes input from readNMEAfiles_lib.py (writes gnssraw.h5) and reads500pingsonardata_wts_gw_ct.py"""

import numpy as np
import matplotlib.pyplot as plt
import tqdm
from scipy import interpolate
import h5py
import pandas as pd
import PostProcessing.yellowfinLib as yellowfinLib
import datetime as DT
from scipy import signal

dirstr = '01-11-2023'
dirstr = '12-21-2022'  # with emlid ppk GPS saved and processed
dirstr = '01-05-2023'
dirstr = '20230327'  # has post processed Emlid data in folder. (*.pos file, maybe multiple)
database = "/home/spike/repos/yellowFin_AcqProcessing/SampleData"
timeString = "20230105"  #""20230417" # "20230327"
datadir = os.path.join(database,timeString)
plotDir = os.path.join(datadir,'figures')
os.makedirs(plotDir, exist_ok=True)  # make folder structure if its not already made
# sonar data
fpathSonar = os.path.join(datadir, 's500') # reads sonar from here
saveFnameSonar = os.path.join(datadir, f'{timeString}_sonarRaw.h5') #saves sonar file here

#NMEA data from sonar, this is not Post Processed Kinematic (PPK) data.  It is used for only cursory or
# introductory look at the data
fpathGNSS = os.path.join(datadir, 'nmeaData')  # load NMEA data from this location
saveFnameGNSS = os.path.join(datadir, f'{timeString}_gnssRaw.h5')  # save nmea data to this location

# RINEX data
# look for all subfolders with RINEX in the folder name inside the "datadir" emlid ppk processor
saveFnamePPK = os.path.join(datadir, f'{timeString}_ppkRaw.h5')

########################################################################################################################
# GPSfname = f"/data/yellowfin/{dirstr}/{dirstr}_gnssRaw.h5" # comes from readNMEAfiles.py
# PPKGPSfname = f"/data/yellowfin/{dirstr}/{dirstr}_ppkRaw.h5" # comes from read_emlid_LLH_raw.py
# S1fname = f"/data/yellowfin/{dirstr}/{dirstr}_sonarRaw.h5" # comes from reads500pingsonardata_wts_gw_ct.py
ET2UTC = 4*60*60 # time in seconds to adjust to UTC from ET (varies depending on time of year)
# load all files
GPS = yellowfinLib.load_h5_to_dictionary(saveFnameGNSS)
ppkGPS = pd.read_hdf(saveFnamePPK)
ppkGPS['epochTime'] = ppkGPS['datetime'].apply(lambda x: x.timestamp()) - 18 # 18 is leap second
# adjustment
ppkGPS['datetime'] = ppkGPS['datetime'] - DT.timedelta(seconds=18) # making sure both are equal
S1 = yellowfinLib.load_h5_to_dictionary(saveFnameSonar)

ppkGPS['GNSS_elevation_NAVD88'] = yellowfinLib.convertEllipsoid2NAVD88(ppkGPS['lat'], ppkGPS['lon'], ppkGPS['height'],
                                                   geoidFile="/home/spike/repos/yellowFin_AcqProcessing/g2012bu0.bin")


# Compare GPS data to make sure timing is ok
plt.figure()
pc_time_off = GPS['pc_time_gga'] + ET2UTC  - GPS['gps_time']
plt.plot(GPS['gps_time'], pc_time_off, '.')
plt.title('time offset between pc time and GPS time')
plt.xlabel('gps time')
plt.ylabel('time offset')

# Use the cerulean instantaneous bed detection since not sure about delay with smoothed
# adjust time of the sonar time stamp with timezone shift (ET -> UTC) and the timeshift between the computer and GPS
S1['time'] = S1['time'] + ET2UTC - np.median(pc_time_off)  # DT.timedelta(hours=5)  # convert to UTC
print('warning yellowfin computer time is not on UTC!!!!!! conversion assumes rigid offset')
sonar_range = S1['smooth_depth_m'] # S1['this_ping_depth_m']
smooth_sonar_range = S1['smooth_depth_m']

# plot ALLLL data
plt.figure(figsize=(16, 4))
plt.plot(S1['time'], sonar_range, 'b.', label='sonar instant depth')
# plt.plot(sonar_time, smooth_sonar_range, 'r.')
plt.plot(GPS['gps_time'], GPS['altMSL'], '.g', label='L1 GPS elev (MSL)')

# gps_leap_offset is
plt.plot(ppkGPS['epochTime'], ppkGPS['GNSS_elevation_NAVD88'], '.r', label='ppk elevation [NAVD88 m]')
plt.ylim([0, 20])
plt.legend()


# plot sonar, select indices of interest, and then second subplot is time of interest
plt.figure(figsize=(18,4))
plt.subplot(211)
plt.title('all data, select start/end point for measured depths\nadd extra time for PPK offset')
plt.plot(sonar_range)
plt.ylim([0,10])
d = plt.ginput(2)

plt.subplot(212)
# Now pull corresponding indices for sonar data for same time
sonarIndicies = np.arange(np.floor(d[0][0]).astype(int), np.ceil(d[1][0]).astype(int))
plt.plot(sonar_range[sonarIndicies])
plt.title('my selected data to proceed with')
plt.tight_layout()
plt.ylim([0,10])

# now identify corresponding times from ppk GPS to those times of sonar that we're interested in
indsPPK = np.where((ppkGPS['epochTime'] >= S1['time'][sonarIndicies[0]]) &
                   (ppkGPS['epochTime'] <= S1['time'][sonarIndicies[-1]]))[0]

## now interpolate the lower sampled (sonar 3.33 hz) to the higher sampled data (gps 10 hz)
# 1. identify common timestamp to interpolate to at higher frequency
commonTime = np.linspace(ppkGPS['epochTime'][indsPPK[0]], ppkGPS['epochTime'][indsPPK[-1]],
                        int((ppkGPS['epochTime'][indsPPK[-1]] - ppkGPS['epochTime'][indsPPK[0]])/.1), endpoint=True)

f = interpolate.interp1d(S1['time'], sonar_range) #, fill_value='extrapolate')
sonar_range_i = f(commonTime)
f = interpolate.interp1d(ppkGPS['epochTime'], ppkGPS['height'])
ppkHeight_i = f(commonTime)
# now i have both signals at the same time stamps
phaseLagInSamps, phaseLaginTime = yellowfinLib.findTimeShiftCrossCorr(signal.detrend(ppkHeight_i),
                                                                      signal.detrend(sonar_range_i),
                                                                      sampleFreq=np.median(np.diff(commonTime)))

# now make figure where i have subset that only looks at sonar elevations and GPS elevations
plt.figure(figsize=(12,8))
plt.subplot(211)
plt.plot(S1['time'][sonarIndicies], sonar_range[sonarIndicies], label='sonar_raw')
plt.plot(ppkGPS['epochTime'][indsPPK], ppkGPS['GNSS_elevation_NAVD88'][indsPPK], label='ppk elevation NAVD88 m')
plt.legend()
plt.subplot(212)
plt.title(f"sonar data needs to be adjusted by {phaseLaginTime} seconds")
plt.plot(commonTime, signal.detrend(ppkHeight_i), label='ppk input')
plt.plot(commonTime, signal.detrend(sonar_range_i), label='sonar input')
plt.plot(commonTime+phaseLaginTime, signal.detrend(sonar_range_i), '.', label='interp _sonar shifted')
plt.legend()
plt.tight_layout()
print(f"sonar data needs to be adjusted by {phaseLaginTime} seconds")

## now process all data for saving to file
antenna_offset = 0.25 # meters between the antenna phase center and sounder head
sonar_time_out = S1['time'] + phaseLaginTime

## ok now put the sonar data on the GNSS timestamps which are decimal seconds.  We can do this with sonar_time_out,
# because we just adjusted by the phase lag to make sure they are timesynced.
timeOutInterpStart =  np.ceil(sonar_time_out.min() * 10)/10 # round to nearest 0.1
timeOutInterpEnd = np.floor(sonar_time_out.max() * 10)/10 # round to nearest 0.1
# create a timestamp for data to be output and in phase with that of the ppk gps data which are on the 0.1 s
time_out = np.linspace(timeOutInterpStart, timeOutInterpEnd, int((timeOutInterpEnd - timeOutInterpStart)/0.1),
                       endpoint=True)

print("here's where some filtering could be done, probably worth saving an intermediate product here for future "
      "revisit")
# # upsample interpolate sonar data to time resolution of gnss
# sonar_SmoothDepth_out = interpolate.interp1d(sonar_time_out, S1['smooth_depth_m'])(time_out)
# sonar_SmoothConfidence_out = interpolate.interp1d(sonar_time_out, S1['smoothed_depth_measurement_confidence'])(time_out)
# sonar_InstantDepth_out = interpolate.interp1d(sonar_time_out, S1['this_ping_depth_m'])(time_out)
# sonar_InstantConfidence_out = interpolate.interp1d(sonar_time_out, S1['this_ping_depth_measurement_confidence'])(time_out)
# sonar_backscatter_out = interpolate.interp2d(sonar_time_out, S1['range_m'], S1['profile_data'])(time_out, S1['range_m'])

# now put relevant GNSS and sonar on output timestamps
#initalize out variables
sonar_smooth_depth_out, sonar_smooth_confidence_out = np.zeros_like(time_out)*np.nan, np.zeros_like(time_out)*np.nan
sonar_instant_depth_out, sonar_instant_confidence_out  = np.zeros_like(time_out)*np.nan, np.zeros_like(time_out)*np.nan
sonar_backscatter_out = np.zeros((time_out.shape[0], S1['range_m'].shape[0]))*np.nan
lat_out, lon_out = np.zeros_like(time_out)*np.nan, np.zeros_like(time_out)*np.nan
elevation_out, fix_quality = np.zeros_like(time_out)*np.nan, np.zeros_like(time_out)*np.nan
for tidx, tt in tqdm.tqdm(enumerate(time_out)):
    idxTimeMatchGNSS, idxTimeMatchGNSS = None, None
    
    #first find if theres a time match for sonar
    sonarlogic = np.abs(np.ceil(tt * 10)/10 - np.ceil(sonar_time_out*10)/10)
    if sonarlogic.min() <= 0.2: # .101 handles weird float numerics
        idxTimeMatchSonar= np.argmin(sonarlogic)
    # then find comparable time match for ppk
    ppklogic = np.abs(np.ceil(tt * 10)/10 - np.ceil(ppkGPS['epochTime'].array*10)/10)
    if ppklogic.min() <= 0.101:  # .101 handles numerics
        idxTimeMatchGNSS = np.argmin(ppklogic)
    # if we have both, then we log the data
    if idxTimeMatchGNSS is not None and idxTimeMatchSonar is not None : # we have matching data
        if ppkGPS['Q'][idxTimeMatchGNSS] <=1 and S1['smoothed_depth_measurement_confidence'][idxTimeMatchSonar] > 60:
            sonar_smooth_depth_out[tidx] = S1['smooth_depth_m'][idxTimeMatchSonar]
            sonar_instant_depth_out[tidx] = S1['this_ping_depth_m'][idxTimeMatchSonar]
            sonar_smooth_confidence_out[tidx] = S1['smoothed_depth_measurement_confidence'][idxTimeMatchSonar]
            sonar_instant_confidence_out[tidx] = S1['this_ping_depth_measurement_confidence'][idxTimeMatchSonar]
            sonar_backscatter_out[tidx] = S1['profile_data'][:, idxTimeMatchSonar]
            lat_out[tidx] = ppkGPS['lat'][idxTimeMatchGNSS]
            lon_out[tidx] = ppkGPS['lon'][idxTimeMatchGNSS]
            elevation_out[tidx] = ppkGPS['GNSS_elevation_NAVD88'][idxTimeMatchGNSS] - antenna_offset - S1['smooth_depth_m'][
                idxTimeMatchSonar]
            fix_quality[tidx] = ppkGPS['Q'][idxTimeMatchGNSS]

# identify data that are not nan's to save
idxDataToSave = np.argwhere(~np.isnan(sonar_smooth_depth_out)) # identify data that are not NaNs

# make a final plot of all the processed data
plt.figure()
plt.scatter(lon_out[idxDataToSave], lat_out[idxDataToSave], c=elevation_out[idxDataToSave])
plt.colorbar()
plt.plot(ppkGPS['lon'], ppkGPS['lat'], 'k.', ms=0.1, alpha=.7)
plt.ylabel('latitude')
plt.xlabel('longitude')
plt.title('final data with elevations')
plt.tight_layout()

outputfile = 'finalDataProduct.h5'
with h5py.File(outputfile, 'w') as hf:
    hf.create_dataset('time', data=time_out[idxDataToSave])
    hf.create_dataset('longitude', data=lon_out[idxDataToSave])
    hf.create_dataset('latitude', data=lat_out[idxDataToSave])
    hf.create_dataset('elevation', data=elevation_out[idxDataToSave])
    hf.create_dataset('fix_quality', data=fix_quality[idxDataToSave])
    hf.create_dataset('sonar_smooth_depth', data=sonar_smooth_depth_out[idxDataToSave])
    hf.create_dataset('sonar_smooth_confidence', data=sonar_smooth_confidence_out[idxDataToSave])
    hf.create_dataset('sonar_instant_depth', data=sonar_instant_depth_out[idxDataToSave])
    hf.create_dataset('sonar_instant_confidence', data=sonar_instant_confidence_out[idxDataToSave])
    hf.create_dataset('sonar_backscatter_out', data=sonar_backscatter_out[idxDataToSave])

File: PostProcessing/MergePPKGPS_sonar.m

% MERGEGPS_SONAR Summary of this function goes here
dirstr='01-11-2023'
dirstr='12-21-2022' % with emlid ppk GPS saved and processed
dirstr='01-05-2023'

GPS=load(['nmea' dirstr '.mat']) % comes from readNMEAfiles.m

PPKGPS=load('PPK20230105141705.mat'); % comes from read_emlid_LLH_raw.m

%% compare gps data to make sure timing is ok
figure(1);clf
plot(GPS.gps_time)
hr=floor(GPS.gps_time(:)/10000);
mm=floor(rem(GPS.gps_time(:)./10000,1).*100);
ss=rem(GPS.gps_time(:)./100,1).*100;
GPS.mat_time=datetime(str2num(dirstr(end-3:end)),str2num(dirstr(1:2)),str2num(dirstr(4:5)),...
    hr,mm,ss);

pc_time_off=GPS.pc_time_gga(1)-GPS.mat_time(1)

plot(GPS.mat_time,GPS.lat,'.')
hold on
gps_leap_offset=seconds(18);%should be 19 but 18 lines it up!
plot(PPKGPS.datetime1-gps_leap_offset,PPKGPS.lat,'r')
plot(PPKGPS.datetime_ppk-gps_leap_offset,PPKGPS.lat_ppk,'g')


%%

if 0 % use the matlab bed deection from profile data
S=load(['profile_data' dirstr '.mat'])
sonar_time=S.dt_profile
sonar_range=S.bed_detect_range';
end

if 1 % use the cerulean instantaenous bed detection since not sure about delay with smoothed
S1=load(['detected_range' dirstr '.mat'])
sonar_time=S1.dt_txt(:)+hours(5);%convert to UTC
sonar_range=S1.txt_depth(:);
end
%%
figure(2);clf
plot(sonar_time,sonar_range,'b')
hold on
plot(GPS.mat_time,GPS.altWGS84,'.g')
hold on
ppk_gpstime=PPKGPS.datetime_ppk-gps_leap_offset;
plot(ppk_gpstime,PPKGPS.height_ell_ppk,'.r')

figure(3);clf
subplot(211)
plot(sonar_range)

subplot(212)
inds=(900:5210)% need to set this manually for each data set
plot(sonar_range(inds))
indsp=near(ppk_gpstime,sonar_time(inds(1))):near(ppk_gpstime,sonar_time(inds(end)));
%%

% code gps
%sonar_range_i=interp1(sonar_time,sonar_range,GPS.pc_time_gga(inds));
%lati=GPS.lat(inds);loni=GPS.lon(inds);

% ppk gps
sonar_range_i=interp1(sonar_time,sonar_range,ppk_gpstime(indsp));
lati=PPKGPS.lat_ppk(indsp);loni=PPKGPS.lon_ppk(indsp);


%% another time sync here by correlating time
figure(40);clf
ws=2100:2400 ;% pick a nice section with a heave effects visble in the echodata and gps

ws2=2000:2500 ;%
% this data doesn't have it so adding artficially
csonar_range_i=sonar_range_i;
csonar_range_i(ws)=csonar_range_i(ws)+.3*sin(ws/10)';

ssr=detrend(csonar_range_i(ws2),'const')+.3*sin(ws2/10)'
hold on
chc=PPKGPS.height_ell_ppk(indsp);
shift=0*pi/4
chc(ws)=(chc(ws)+.3*sin(ws/10+shift)');

sdhc=detrend(chc(ws2)+.3*sin(ws2/10+shift)','const')
subplot(412)
plot(sdhc);
hold on
plot(ssr);
subplot(411)
plot(csonar_range_i)
hold on
plot(chc)


subplot(413)
[r,lags]=xcorr(ssr,sdhc,200,'coeff');
plot(lags,r)

[mx,mi]=max(r);
ml=lags(mi)
s_ssr=interp1(1:length(ssr),ssr,[1:length(ssr) ]+ml);
shifted_sonar_range_i=interp1(1:length(sonar_range_i),sonar_range_i,[1:length(sonar_range_i) ]+ml);

subplot(414)
plot(sdhc);
hold on
plot(s_ssr);
%%
figure(4) ;clf
scatter(loni,lati,16,sonar_range_i)
cc=turbo;
cc=flipud(cc);
colormap(cc)
colorbar
hold on
caxis([7 13])

%%
antenna_offset=.25;
x=loni(:);
y=lati(:);
z=-sonar_range_i(:)+PPKGPS.height_ell_ppk(indsp)-antenna_offset;
dx=.000005;
xnodes=(min(x)-dx):dx:(max(x)+dx);
ynodes=(min(y)-dx):dx:(max(y)+dx);
sff= 2e-04;
[Zg3,Xg3,Yg3]=RegularizeData3D(x ,y,z,xnodes,ynodes,'smoothness',sff);
% RegularizeData3D needs to be downloaded from the matlab file exchange
%%
figure(22);clf
scatter3(x,y,z,12,z,'filled');
view(2)
hold on
% hold on
% sl=600
% for ii=1:(floor(length(x)/sl)-3)
%     fit_inds=[1:sl]+sl+sl*ii
%     scatter(x(fit_inds),y(fit_inds),12,z(fit_inds),'or');
%     pause(.1)
% end
k=boundary(x,y,.3);%x and y are track line coordinate vectors in UTM, z is the surface re navd88
plot(x(k),y(k),'k','linewidth',3)
[m,n]=size(Zg3)
BW=roipoly(Xg3(1,:),Yg3(:,1),Zg3,x(k),y(k));
BW=double(BW);
BW(BW==0)=NaN;


DEM=GRIDobj(Xg3,Yg3,Zg3.*BW);
set(gca,'dataaspectratio',[1 1 1])


%%
figure(31);clf
cm=(ttscm('broc'));
%cm=ttcmap('etopo1')
wgsz=32
  dataMax = -1-wgsz;
  dataMin = -14-wgsz;
  centerPoint = -10-wgsz;
  scalingIntensity = 4;
%Then perform some operations to create your colormap. I have done this by altering the indices “x” at which each existing color lives, and then interpolating to expand or shrink certain areas of the spectrum.
  xm = 1:length(cm);
  xm = xm - (centerPoint-dataMin)*length(xm)/(dataMax-dataMin);
  xm = scalingIntensity * xm/max(abs(xm));
%Next, select some function or operations to transform the original linear indices into nonlinear. In the last line, I then use “interp1” to create the new colormap from the original colormap and the transformed indices.
  xm = sign(xm).* exp(abs(xm));
  xm = xm - min(xm); xm = xm*511/max(xm)+1;
  ncm = interp1(xm, cm, 1:512);

%contour(DEM,[],'caxis',[-18 -7],'colorbar',1,'colormap',cm,'ticklabels','nice','exaggerate',1,...
 %   'colorbarylabel','Elevation (M)')
tanakacontour(DEM,[-14:.25:-2]-wgsz)
%caxis([-14 -3]-wgsz)
title('Yellow Fin survey')
hold on
%plot(x,y,'.k');
sf =10;
hs=scatter(x(1:sf:end),y(1:sf:end),12,z(1:sf:end),'o')
hs.MarkerEdgeColor='k'
%cc=flipud(cc);
colormap(ncm)
hc=colorbar
hc.Label.String='Depth (m)'
xlabel('Lon (Deg)')
ylabel('Lat (Deg)')
print -dpng goodwill_pond